File: tools/js-ts-indexer/src/openapi-pathrefs.ts
/**
 * In-document OpenAPI $ref resolution for Path Item and Operation objects (JSON Pointer, RFC 6901).
 * External references (no leading `#`) are not followed — keeps indexing deterministic without I/O.
 */

const MAX_REF_DEPTH = 24;

function jsonPointerResolve(doc: unknown, ref: string): unknown {
  let pointer = ref.startsWith("#") ? ref.slice(1) : ref;
  if (pointer === "") return doc;
  if (!pointer.startsWith("/")) return undefined;
  const segments = pointer.slice(1).split("/");
  let cur: unknown = doc;
  for (const raw of segments) {
    const tok = raw.replace(/~1/g, "/").replace(/~0/g, "~");
    if (cur === null || typeof cur !== "object") return undefined;
    if (Array.isArray(cur)) {
      const idx = parseInt(tok, 10);
      if (Number.isNaN(idx) || idx < 0 || idx >= cur.length) return undefined;
      cur = cur[idx];
    } else {
      const o = cur as Record<string, unknown>;
      if (!Object.prototype.hasOwnProperty.call(o, tok)) return undefined;
      cur = o[tok];
    }
  }
  return cur;
}

function shallowMergeOpenAPI(base: Record<string, unknown>, overlay: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = { ...base };
  for (const [k, v] of Object.entries(overlay)) {
    if (k === "$ref") continue;
    out[k] = v;
  }
  return out;
}

function resolveOpenAPIOperation(
  op: Record<string, unknown>,
  root: Record<string, unknown>,
  depth: number,
  seen: Set<string>,
): Record<string, unknown> {
  let m = shallowMergeOpenAPI(op, {});
  while (depth < MAX_REF_DEPTH) {
    const ref = m.$ref;
    if (typeof ref !== "string" || !ref.startsWith("#")) break;
    if (seen.has(ref)) break;
    seen.add(ref);
    const target = jsonPointerResolve(root, ref);
    seen.delete(ref);
    if (!target || typeof target !== "object" || Array.isArray(target)) break;
    m = shallowMergeOpenAPI(target as Record<string, unknown>, m);
    delete m.$ref;
    depth++;
  }
  return m;
}

function resolveOpenAPIPathItem(val: unknown, root: Record<string, unknown>, depth: number, seen: Set<string>): Record<string, unknown> | null {
  if (!val || typeof val !== "object" || Array.isArray(val)) return null;
  let m = shallowMergeOpenAPI(val as Record<string, unknown>, {});
  while (depth < MAX_REF_DEPTH) {
    const ref = m.$ref;
    if (typeof ref !== "string" || !ref.startsWith("#")) break;
    if (seen.has(ref)) break;
    seen.add(ref);
    const target = jsonPointerResolve(root, ref);
    seen.delete(ref);
    if (!target || typeof target !== "object" || Array.isArray(target)) break;
    m = shallowMergeOpenAPI(target as Record<string, unknown>, m);
    delete m.$ref;
    depth++;
  }
  for (const vk of Object.keys(m)) {
    const vlow = vk.toLowerCase().trim();
    if (!HTTP_VERB_SET.has(vlow)) continue;
    const vv = m[vk];
    if (vv && typeof vv === "object" && !Array.isArray(vv)) {
      m[vk] = resolveOpenAPIOperation(vv as Record<string, unknown>, root, depth + 1, seen);
    }
  }
  return m;
}

/** Set of HTTP methods (lowercase) — must match openapi-routes HTTP_VERBS iteration. */
const HTTP_VERB_SET = new Set(["get", "post", "put", "patch", "delete", "options", "head", "trace"]);

export function expandOpenAPIPathRefs(pathsVal: unknown, root: Record<string, unknown>): unknown {
  if (!pathsVal || typeof pathsVal !== "object" || Array.isArray(pathsVal)) {
    return pathsVal;
  }
  const pathsObj = pathsVal as Record<string, unknown>;
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(pathsObj)) {
    const resolved = resolveOpenAPIPathItem(v, root, 0, new Set<string>());
    out[k] = resolved ?? v;
  }
  return out;
}

File: internal/intelligence/indexer/openapi_pathrefs.go
package indexer

import (
	"strconv"
	"strings"
)

// openAPIResolveInternalRefs expands in-document $ref on Path Item and Operation objects (OpenAPI 3.x /
// Swagger 2 style) using JSON Pointer (RFC 6901). External refs (no leading #) are left unresolved.
// Cycles and depth over maxOpenAPIRefDepth are capped to keep indexing robust.

const maxOpenAPIRefDepth = 24

func jsonPointerResolve(doc interface{}, ref string) interface{} {
	pointer := ref
	if strings.HasPrefix(pointer, "#") {
		pointer = pointer[1:]
	}
	if pointer == "" {
		return doc
	}
	if !strings.HasPrefix(pointer, "/") {
		return nil
	}
	parts := strings.Split(pointer, "/")
	if len(parts) > 0 && parts[0] == "" {
		parts = parts[1:]
	}
	cur := doc
	for _, raw := range parts {
		tok := strings.ReplaceAll(strings.ReplaceAll(raw, "~1", "/"), "~0", "~")
		switch v := cur.(type) {
		case map[string]interface{}:
			nx, ok := v[tok]
			if !ok {
				return nil
			}
			cur = nx
		case []interface{}:
			idx, err := strconv.Atoi(tok)
			if err != nil || idx < 0 || idx >= len(v) {
				return nil
			}
			cur = v[idx]
		default:
			return nil
		}
	}
	return cur
}

func shallowMergeOpenAPI(base, overlay map[string]interface{}) map[string]interface{} {
	out := make(map[string]interface{}, len(base)+len(overlay))
	for k, v := range base {
		out[k] = v
	}
	for k, v := range overlay {
		if k == "$ref" {
			continue
		}
		out[k] = v
	}
	return out
}

func resolveOpenAPIPathItem(val interface{}, root map[string]interface{}, depth int, seen map[string]bool) map[string]interface{} {
	m, ok := val.(map[string]interface{})
	if !ok {
		return nil
	}
	m = shallowMergeOpenAPI(m, map[string]interface{}{})
	for depth < maxOpenAPIRefDepth {
		ref, ok := m["$ref"].(string)
		if !ok || !strings.HasPrefix(ref, "#") {
			break
		}
		if seen[ref] {
			break
		}
		seen[ref] = true
		target := jsonPointerResolve(root, ref)
		seen[ref] = false
		tm, ok := target.(map[string]interface{})
		if !ok {
			break
		}
		m = shallowMergeOpenAPI(tm, m)
		delete(m, "$ref")
		depth++
	}
	for vk, vv := range m {
		vlow := strings.ToLower(strings.TrimSpace(vk))
		if _, isVerb := openapiHTTPVerbs[vlow]; !isVerb {
			continue
		}
		if om, ok := vv.(map[string]interface{}); ok {
			m[vk] = resolveOpenAPIOperation(om, root, depth+1, seen)
		}
	}
	return m
}

func resolveOpenAPIOperation(op map[string]interface{}, root map[string]interface{}, depth int, seen map[string]bool) map[string]interface{} {
	m := shallowMergeOpenAPI(op, map[string]interface{}{})
	for depth < maxOpenAPIRefDepth {
		ref, ok := m["$ref"].(string)
		if !ok || !strings.HasPrefix(ref, "#") {
			break
		}
		if seen[ref] {
			break
		}
		seen[ref] = true
		target := jsonPointerResolve(root, ref)
		seen[ref] = false
		tm, ok := target.(map[string]interface{})
		if !ok {
			break
		}
		m = shallowMergeOpenAPI(tm, m)
		delete(m, "$ref")
		depth++
	}
	return m
}

func expandOpenAPIPathRefs(paths map[string]interface{}, root map[string]interface{}) map[string]interface{} {
	if paths == nil {
		return nil
	}
	out := make(map[string]interface{}, len(paths))
	for k, v := range paths {
		resolved := resolveOpenAPIPathItem(v, root, 0, make(map[string]bool))
		if resolved != nil {
			out[k] = resolved
		} else {
			out[k] = v
		}
	}
	return out
}

func openapiPathsMapToOps(paths map[string]interface{}) []openapiOp {
	if paths == nil {
		return nil
	}
	var out []openapiOp
	for pth, item := range paths {
		if pth == "" || !strings.HasPrefix(pth, "/") {
			continue
		}
		methods, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		for verb, opVal := range methods {
			vlow := strings.ToLower(strings.TrimSpace(verb))
			if _, ok := openapiHTTPVerbs[vlow]; !ok {
				continue
			}
			opID := ""
			if opMap, ok := opVal.(map[string]interface{}); ok {
				if oid, ok := opMap["operationId"].(string); ok {
					opID = strings.TrimSpace(oid)
				}
			}
			out = append(out, openapiOp{
				method:      strings.ToUpper(vlow),
				path:        pth,
				operationID: opID,
			})
		}
	}
	return out
}
